node/core/pvf/worker/src/execute.rs

```diff
@@ -89,31 +96,57 @@ pub fn worker_entrypoint(socket_path: &str, node_version: Option<&str>) {
 				artifact_path.display(),
 			);
 
-			// Used to signal to the cpu time monitor thread that it can finish.
-			let (finished_tx, finished_rx) = channel::<()>();
+			// Conditional variable to notify us when a thread is done.
+			let cond_main = Arc::new((Mutex::new(true), Condvar::new()));
+			let cond_cpu = Arc::clone(&cond_main);
+			let cond_job = Arc::clone(&cond_main);
+
 			let cpu_time_start = ProcessTime::now();
 
 			// Spawn a new thread that runs the CPU time monitor.
-			let cpu_time_monitor_fut = rt_handle
-				.spawn_blocking(move || {
-					cpu_time_monitor_loop(cpu_time_start, execution_timeout, finished_rx)
-				})
-				.fuse();
+			let (cpu_time_monitor_tx, cpu_time_monitor_rx) = channel::<()>();
+			let cpu_time_monitor_thread = thread::spawn(move || {
+				let result =
+					cpu_time_monitor_loop(cpu_time_start, execution_timeout, cpu_time_monitor_rx);
+				cond_notify_all(cond_cpu);
+				result
+			});
 			let executor_2 = executor.clone();
-			let execute_fut = rt_handle
-				.spawn_blocking(move || {
-					validate_using_artifact(&artifact_path, &params, executor_2, cpu_time_start)
-				})
-				.fuse();
-
-			pin_mut!(cpu_time_monitor_fut);
-			pin_mut!(execute_fut);
-
-			let response = select_biased! {
-				// If this future is not selected, the join handle is dropped and the thread will
-				// finish in the background.
-				cpu_time_monitor_res = cpu_time_monitor_fut => {
-					match cpu_time_monitor_res {
+			let execute_thread =
+				thread::Builder::new().stack_size(EXECUTE_THREAD_STACK_SIZE).spawn(move || {
+					let result = validate_using_artifact(
+						&artifact_path,
+						&params,
+						executor_2,
+						cpu_time_start,
+					);
+					cond_notify_all(cond_job);
+					result
+				})?;
+
+			// Wait for one of the threads to finish.
+			cond_wait_while(cond_main);
+
+			// A thread has signaled completion but it may not be "joinable" yet, so loop for a bit.
+			let response = loop {
+				// NOTE: We check the worker thread first to give it priority. This is for the rare
```

Comment:
This priority check again introduces a race condition that you even documented, when the waiting thread was notified by cpu monitor, but instead checks the execution thread.

You have a mutexed value associated with a condition variable.  For example, store
```rust
enum ExecutionOutcome {
    Finished,
    CpuTimedOut,
    Pending,
}

impl ExecutionOutcome {
    pub fn is_pending(&self) -> bool { /* ... */ }
}
```
 instead of boolean.
Once the main thread is notified, **join** the thread w.r.t. to outcome.
The whole point of it is there's no loop involved at all.

The only trick is that if there's a panic, you don't get any notification. I don't remember if host's worker pool does anything about it, but you can also time out here with some huge value.

Reply:
Done, let me know if that's better and I'll fix up the API a bit to make it nicer to read.

About panics, we should handle them properly, see `cond_notify_on_done`.